iothub/device/src/Pipeline/ConnectionStateDelegatingHandler.cs

```diff
@@ -0,0 +1,382 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.Transport
+{
+    internal sealed class ConnectionStateDelegatingHandler : DefaultDelegatingHandler
+    {
+
+        private readonly SemaphoreSlim _clientOpenSemaphore = new(1, 1);
+
+        private readonly Action<ConnectionStatusInfo> _onConnectionStatusChanged;
+        private readonly IIotHubClientRetryPolicy _retryPolicy;
+
+        private readonly ClientTransportStateMachine _clientTransportStateMachine = new();
+        private Task _transportClosedTask;
+        private CancellationTokenSource _handleDisconnectCts;
+        private CancellationTokenSource _cancelPendingOperationsCts;
+
+        internal ConnectionStateDelegatingHandler(PipelineContext context, IDelegatingHandler innerHandler)
+            : base(context, innerHandler)
+        {
+            _onConnectionStatusChanged = context.ConnectionStatusChangeHandler;
+
+            // This retry policy is saved only to handle the client state when the retry policy informs that no more retry attempts are to be made
+            _retryPolicy = context.RetryPolicy;
+        }
+
+        public override async Task OpenAsync(CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, cancellationToken, nameof(OpenAsync));
+
+            try
+            {
+                switch (_clientTransportStateMachine.GetCurrentState())
+                {
+                    case ClientTransportState.Open:
+                        return;
+
+                    case ClientTransportState.Closing:
+                        throw new InvalidOperationException($"The client is currently closing. To reopen the client wait until {nameof(CloseAsync)} completes" +
+                            $" and then invoke {nameof(OpenAsync)}.");
+
+                    case ClientTransportState.Opening:
+                    case ClientTransportState.Closed:
+                        {
```

Comment:
This block is big enough, it could warrant being in a private method to improve readability of this method.

Reply:
On second thought, I think I'm going to leave this block as it is. If I split it out to its own method then we'll need to add documentation &/or checks to check for the other switch cases that we have above so that it doesn't execute in cases that it shouldn't. 
Since it is expected to be called from only a single location, it doesn't seem worth it to split it out.

Reply:
I think you could document on the private method that it should only be called from one place, and you get better readability out of it. But that's just my opinion and current take.